// @uifabric/styling/lib-amd/examples/AnimationTile/AnimationTile.styles.js
define(["require", "exports", "tslib", "@uifabric/styling"], function (require, exports, tslib_1, styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    function getStyles(theme) {
        if (theme === void 0) { theme = styling_1.getTheme(); }
        return {
            root: styling_1.mergeStyles({
                marginBottom: '20px'
            }),
            title: styling_1.mergeStyles(tslib_1.__assign({}, theme.fonts.medium, { marginBottom: '8px' })),
            container: styling_1.mergeStyles({
                overflow: 'hidden',
                position: 'relative',
                maxWidth: '400px',
                height: '100px',
                border: '1px solid black',
                backgroundImage: 'url("data:image/svg+xml;base64,' + 'PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHdpZHRoPScxMCcgaGVpZ2h0PScxMCc' +
                    '+CiAgPHJlY3Qgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJyBmaWxsPSd3aGl0ZScvPgogIDxwYXRoIGQ9J00tMSwx' +
                    'IGwyLC0yCiAgICAgICAgICAgTTAsMTAgbDEwLC0xMAogICAgICAgICAgIE05LDExIGwyLC0yJyBzdHJva2U9J' +
                    '2JsYWNrJyBzdHJva2Utd2lkdGg9JzEnLz4KPC9zdmc+Cg==")',
                backgroundRepeat: 'repeat',
            }),
            animationBox: styling_1.mergeStyles({
                position: 'absolute',
                background: theme.palette.themePrimary,
                width: '25%',
                height: '100%'
            }),
            isLeft: styling_1.mergeStyles({
                left: 0,
                top: 0,
                width: '25%',
                height: '100%'
            }),
            isRight: styling_1.mergeStyles({
                right: 0,
                top: 0,
                width: '25%',
                height: '100%'
            }),
            isTop: styling_1.mergeStyles({
                left: 0,
                top: 0,
                width: '100%',
                height: '50%'
            }),
            isBottom: styling_1.mergeStyles({
                left: 0,
                bottom: 0,
                width: '100%',
                height: '50%'
            }),
            isIn: styling_1.mergeStyles({
                opacity: 0
            })
        };
    }
    exports.getStyles = getStyles;
});
