// @uifabric/styling/lib-amd/examples/ColorPage/ColorPage.styles.js
define(["require", "exports", "@uifabric/styling"], function (require, exports, styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    function getStyles(theme) {
        if (theme === void 0) { theme = styling_1.getTheme(); }
        return {
            swatch: styling_1.mergeStyles({
                display: 'inline-block',
                verticalAlign: 'middle',
                width: 20,
                height: 20,
                borderRadius: '50%',
                marginRight: '8px'
            }),
            cell: styling_1.mergeStyles(theme.fonts.small, {
                display: 'inline-block',
                verticalAlign: 'middle',
            }),
            cellWithBorder: styling_1.mergeStyles({
                borderWidth: '1px',
                borderStyle: 'solid'
            }),
            swatchText: styling_1.mergeStyles({
                display: 'inline-block',
                verticalAlign: 'middle',
            }),
            example: styling_1.mergeStyles(theme.fonts.mediumPlus)
        };
    }
    exports.getStyles = getStyles;
});
