// office-ui-fabric-react/lib-amd/components/HoverCard/ExpandingCard.styles.js
define(["require", "exports", "../../Utilities", "../../Styling"], function (require, exports, Utilities_1, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = Utilities_1.memoizeFunction(function (theme, customStyles) {
        var styles = {
            root: {
                width: '340px',
                'pointer-events': 'none',
                '.ms-Callout': {
                    'box-shadow': '0 0 20px rgba(0, 0, 0, .2)',
                    border: 'none'
                }
            },
            compactCard: [
                {
                    'pointer-events': 'auto',
                    position: 'relative'
                },
                Styling_1.after({
                    content: '""',
                    position: 'absolute',
                    bottom: '0',
                    left: '0',
                    right: '0',
                    height: '1px',
                    'background-color': theme.palette.neutralLighter
                })
            ],
            expandedCard: {
                height: '0',
                'overflow-y': 'hidden',
                'pointer-events': 'auto',
                transition: 'height 0.467s cubic-bezier(0.5, 0, 0, 1)'
            }
        };
        return Styling_1.mergeStyleSets(styles, customStyles);
    });
});
