// office-ui-fabric-react/lib-amd/components/CommandBar/examples/CommandBar.Example.scss.js
define(["require", "exports", "@microsoft/load-themed-styles"], function (require, exports, load_themed_styles_1) {
    "use strict";
    var styles = {
        customButtonContainer: 'customButtonContainer_fdece0ef',
        icon: 'icon_fdece0ef',
        themeDarkAltColor: 'themeDarkAltColor_fdece0ef',
        leftText: 'leftText_fdece0ef',
        splitter: 'splitter_fdece0ef',
        button: 'button_fdece0ef',
        darkerBG: 'darkerBG_fdece0ef',
    };
    load_themed_styles_1.loadStyles([{ "rawString": ".customButtonContainer_fdece0ef{display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-align:center;-ms-flex-align:center;align-items:center}.icon_fdece0ef{display:inline-block;font-family:FabricMDL2Icons;font-style:normal;font-weight:400;speak:none;font-size:14px;position:relative}.themeDarkAltColor_fdece0ef{color:" }, { "theme": "themeDarkAlt", "defaultValue": "#106ebe" }, { "rawString": "}.leftText_fdece0ef{margin-left:6px;font-size:14px;font-weight:400;color:" }, { "theme": "neutralPrimary", "defaultValue": "#333333" }, { "rawString": "}.splitter_fdece0ef{font-size:14px;font-weight:400;color:" }, { "theme": "neutralTertiary", "defaultValue": "#a6a6a6" }, { "rawString": ";background:0 0;position:relative;top:-2px;height:40px}.button_fdece0ef.ms-Button{height:40px;min-width:28px;padding:0 5px}.button_fdece0ef:hover{background:initial}.darkerBG_fdece0ef{background:#d0d0d0}" }]);
    return styles;
});
/* tslint:enable */
